pkg/xds/topology/outbound.go

```diff
@@ -72,63 +119,136 @@ func BuildEdsEndpointMap(
 		if !zi.IsRemoteIngress(zone) {
 			continue
 		}
+
 		if !mesh.MTLSEnabled() {
-			// Ingress routes the request by TLS SNI, therefore for cross cluster communication MTLS is required
-			// We ignore Ingress from endpoints if MTLS is disabled, otherwise we would fail anyway.
+			// Ingress routes the request by TLS SNI, therefore for cross
+			// cluster communication MTLS is required.
+			// We ignore Ingress from endpoints if MTLS is disabled, otherwise
+			// we would fail anyway.
 			continue
 		}
+
 		if !zi.HasPublicAddress() {
-			continue // Zone Ingress is not reachable yet from other clusters. This may happen when Ingress Service is pending waiting on External IP on Kubernetes.
+			// Zone Ingress is not reachable yet from other clusters.
+			// This may happen when Ingress Service is pending waiting on
+			// External IP on Kubernetes.
+			continue
 		}
-		ingressCoordinates := net.JoinHostPort(zi.Spec.GetNetworking().GetAdvertisedAddress(), strconv.FormatUint(uint64(zi.Spec.GetNetworking().GetAdvertisedPort()), 10))
-		if ingressInstances[ingressCoordinates] {
-			continue // many Ingress instances can be placed in front of one load balancer (all instances can have the same public address and port). In this case we only need one Instance avoiding creating unnecessary duplicated endpoints
+
+		ziNetworking := zi.Spec.GetNetworking()
+		ziAddress := ziNetworking.GetAdvertisedAddress()
+		ziPort := ziNetworking.GetAdvertisedPort()
+		ziCoordinates := buildCoordinates(ziAddress, ziPort)
+
+		if _, ok := ziInstances[ziCoordinates]; ok {
+			// many Ingress instances can be placed in front of one load
+			// balancer (all instances can have the same public address and
+			// port).
+			// In this case we only need one Instance avoiding creating
+			// unnecessary duplicated endpoints
+			continue
 		}
-		ingressInstances[ingressCoordinates] = true
+
+		ziInstances[ziCoordinates] = struct{}{}
+
 		for _, service := range zi.Spec.GetAvailableServices() {
 			if service.Mesh != mesh.GetMeta().GetName() {
 				continue
 			}
-			serviceName := service.Tags[mesh_proto.ServiceTag]
-			outbound[serviceName] = append(outbound[serviceName], core_xds.Endpoint{
-				Target:   zi.Spec.GetNetworking().GetAdvertisedAddress(),
-				Port:     zi.Spec.GetNetworking().GetAdvertisedPort(),
-				Tags:     service.Tags,
-				Weight:   service.Instances,
-				Locality: localityFromTags(mesh, priorityRemote, service.Tags),
-			})
+
+			serviceTags := service.GetTags()
+			serviceName := serviceTags[mesh_proto.ServiceTag]
+			serviceInstances := service.GetInstances()
+			locality := localityFromTags(mesh, priorityRemote, serviceTags)
+
+			// TODO (bartsmykla): We have to check if it will be ok in a situation
+			//  where we have few zone ingresses with the same services, as
+			//  with zone egresses we will generate endpoints with the same
+			//  targets and ports (zone egress ones), which envoy probably will
+			//  ignore
```

Comment:
Please make a github issue (or todo list after merge if you plan to check it) so we don't forget to investigate it.